### PR TITLE
BUG: misleading error creating df from 2d array

### LINKED ISSUE #42463: BUG: Constructing a df using 2D numpy arrays raises misleading error
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
>>> import pandas as pd
>>> import numpy as np
>>> a = [[1], [2]]
>>> pd.DataFrame({'a': np.array(a), 'b': np.array(a)}) 
ValueError: If using all scalar values, you must pass an index
```

#### Problem description

Hey all! Thanks for the library, I use pandas on a daily basis.

When attempting to create a DataFrame with a 2-dimensional numpy array as input, a misleading `ValueError: If using all scalar values, you must pass an index` is raised instead of raising an expected `ValueError: Data must be 1-dimensional`.

This behaviour changes slightly if there is more than one array:
```python
>>> import pandas as pd
>>> import numpy as np
>>> a = [[1], [2]]
>>> pd.DataFrame({'a': a, 'b': a})   # No error, both nested lists pass validation and index is extracted
     a    b
0  [1]  [1]
1  [2]  [2]
>>> pd.DataFrame({'a': np.array(a), 'b': a})   # Expected behaviour, the nested list passes the check at pd.core.internals.construction._extract_index
ValueError: Data must be 1-dimensional
>>> pd.DataFrame({'a': np.array(a), 'b': np.array(a)}) 
ValueError: If using all scalar values, you must pass an index
```

I suspect this may have something to do with the check at https://github.com/pandas-dev/pandas/blob/386bad4c900ed09c1b664a341e3081851694ddbc/pandas/core/internals/construction.py#L620 
not being exhaustive and/or that the shape validation only happens after this index creation step.

#### Expected Output

```python
>>> pd.DataFrame({'a': np.array(a), 'b': np.array(a)}) 
ValueError: Data must be 1-dimensional
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : f00ed8f47020034e752baf0250483053340971b0
python           : 3.7.10.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.104+
Version          : #1 SMP Sat Jun 5 09:50:34 PDT 2021
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.0
numpy            : 1.19.5
pytz             : 2018.9
dateutil         : 2.8.1
pip              : 19.3.1
setuptools       : 57.0.0
Cython           : 0.29.23
pytest           : 3.6.4
hypothesis       : None
sphinx           : 1.8.5
blosc            : None
feather          : 0.4.1
xlsxwriter       : None
lxml.etree       : 4.2.6
html5lib         : 1.0.1
pymysql          : None
psycopg2         : 2.7.6.1 (dt dec pq3 ext lo64)
jinja2           : 2.11.3
IPython          : 5.5.0
pandas_datareader: 0.9.0
bs4              : 4.6.3
bottleneck       : 1.3.2
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.2.2
numexpr          : 2.7.3
odfpy            : None
openpyxl         : 2.5.9
pandas_gbq       : 0.13.3
pyarrow          : 3.0.0
pyxlsb           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : 1.4.20
tables           : 3.4.4
tabulate         : 0.8.9
xarray           : 0.18.2
xlrd             : 1.1.0
xlwt             : 1.3.0
numba            : 0.51.2

</details>

### PR DESCRIPTION
**Pull request by debnathshoham**
- [x] closes #42463
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/internals/construction.py [hunk 1]
```diff
@@ -615,6 +615,8 @@ def _extract_index(data) -> Index:
             elif is_list_like(val) and getattr(val, "ndim", 1) == 1:
                 have_raw_arrays = True
                 raw_lengths.append(len(val))
+            elif isinstance(val, np.ndarray) and getattr(val, "ndim", 1) > 1:
```
Comment (jbrockmendel) [inline pandas/core/internals/construction.py:618]: after the isinstance check the getattr can check can just be `val.ndim > 1`

Comment (debnathshoham) [reply pandas/core/internals/construction.py]: amended

**Comment by debnathshoham** [comment]
`FAILED pandas/core/groupby/groupby.py::pandas.core.groupby.groupby.GroupBy.apply`
Failure unrelated to my change

File: pandas/core/internals/construction.py [hunk 2]
```diff
@@ -615,6 +615,8 @@ def _extract_index(data) -> Index:
             elif is_list_like(val) and getattr(val, "ndim", 1) == 1:
                 have_raw_arrays = True
                 raw_lengths.append(len(val))
+            elif isinstance(val, np.ndarray) and val.ndim > 1:
+                raise ValueError("Data must be 1-dimensional")
```
Comment (jbrockmendel) [inline pandas/core/internals/construction.py:619]: maybe something like "Per-column arrays must each be 1-dimensional"?

Comment (debnathshoham) [reply pandas/core/internals/construction.py]: amended

File: pandas/tests/frame/test_constructors.py [hunk 3]
```diff
@@ -2530,6 +2530,19 @@ def test_from_2d_object_array_of_periods_or_intervals(self):
...
+    @pytest.mark.parametrize(
+        "col_a, col_b",
+        [
+            ([[1], [2]], np.array([[1], [2]])),
+            (np.array([[1], [2]]), [[1], [2]]),
+            (np.array([[1], [2]]), np.array([[1], [2]])),
+        ],
+    )
+    def test_error_from_2darray(self, col_a, col_b):
+        msg = "Data must be 1-dimensional"
```
Comment (jbrockmendel) [inline pandas/tests/frame/test_constructors.py:2542]: probably need to update the message here?

Comment (debnathshoham) [reply pandas/tests/frame/test_constructors.py]: My bad. Updated the msg in test.

**Review by jbrockmendel — APPROVED** [review]
LGTM

**Comment by jreback** [comment]
@debnathshoham lgtm. can you add a whatsnew note (bug fixes for reshaping for 1.4 is good), a short note about the revised error that a user would see. ping on green.

**Comment by debnathshoham** [comment]
@jreback Green-ish (unrelated failures)

**Comment by jreback** [comment]
thanks @debnathshoham

**Pull request merged**